worker/uniter/resolver.go

```diff
@@ -260,7 +260,8 @@ func (s *uniterResolver) nextOp(
 		return opFactory.NewUpgrade(remoteState.CharmURL)
 	}
 
-	if localState.ConfigVersion != remoteState.ConfigVersion {
+	if localState.ConfigVersion != remoteState.ConfigVersion ||
+		localState.Series != remoteState.Series {
```

Comment:
This probably isn't sufficient since exiting local state will not have series set at all. We could argue that config-changed hooks should be idempotent and so running an extra one won't matter, but it sort of does at scale. Now, after a Juju upgrade, we may well run a config-changed anyway when the agent restarts, and the hook commit would update the series then. Can we check to see if that's the behaviour? The other option is to implement a uniter upgrade step to update all the local states.

Reply:
Verified: when the agent is upgraded, a restart is required, which causes most (all) hooks to be rerun including config-changed.  Spurious config-changed hooks are not running after an upgrade to an agent with this functionality.